### PR TITLE
[FLINK-18850][Table SQL / Runtime]Add late records dropped metric for row time over windows

### PR DESCRIPTION
**Pull request by zhuxiaoshang**
## What is the purpose of the change

Currently all the row time over windows in blink planner runtime discards late records silently, it would be good to have a metric about the late records dropping.

## Brief change log

- add metrics of late records dropping to all row time over windows

## Verifying this change

This change added tests and can be verified as follows:
- RowTimeRangeBoundedPrecedingFunctionTest
- RowTimeRangeUnboundedPrecedingFunctionTest
- RowTimeRowsBoundedPrecedingFunctionTest
- RowTimeRowsUnboundedPrecedingFunctionTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (no)

### CODE REVIEW

**Comment by zhuxiaoshang** [comment]
@flinkbot run azure

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/over/RowTimeRangeBoundedPrecedingFunctionTest.java [hunk 1]
```diff
@@ -40,25 +33,7 @@
...
-	private LogicalType[] inputFieldTypes = new LogicalType[]{
-		new VarCharType(VarCharType.MAX_LENGTH),
-		new BigIntType(),
-		new BigIntType()
-	};
-	private LogicalType[] accTypes = new LogicalType[]{ new BigIntType() };
-
-	private BinaryRowDataKeySelector keySelector = new BinaryRowDataKeySelector(new int[]{ 0 }, inputFieldTypes);
-	private TypeInformation<RowData> keyType = keySelector.getProducedType();
+public class RowTimeRangeBoundedPrecedingFunctionTest extends RowTimeOverWindowTestBase{
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/over/RowTimeRangeBoundedPrecedingFunctionTest.java:36]:
```suggestion
public class RowTimeRangeBoundedPrecedingFunctionTest extends RowTimeOverWindowTestBase {
```

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/over/RowTimeRangeUnboundedPrecedingFunctionTest.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+import org.junit.Test;
+
+import static org.apache.flink.table.runtime.util.StreamRecordUtils.insertRecord;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Test for {@link RowTimeRangeUnboundedPrecedingFunction}.
+ */
+public class RowTimeRangeUnboundedPrecedingFunctionTest extends RowTimeOverWindowTestBase{
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/over/RowTimeRangeUnboundedPrecedingFunctionTest.java:35]:
```suggestion
public class RowTimeRangeUnboundedPrecedingFunctionTest extends RowTimeOverWindowTestBase {
```

**Review by libenchao — CHANGES_REQUESTED** [review]
@zhuxiaoshang Thanks for your contribution, the PR LGTM generally. I left some comments.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/RowTimeRangeBoundedPrecedingFunction.java [hunk 3]
```diff
@@ -86,12 +90,25 @@
 
 	private transient AggsHandleFunction function;
 
+	// ------------------------------------------------------------------------
+	// Metrics
+	// ------------------------------------------------------------------------
+	protected static final String LATE_ELEMENTS_DROPPED_METRIC_NAME = "numLateRecordsDropped";
+	protected static final String LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME = "lateRecordsDroppedRate";
```
[the remark this answers is not in the document]
Comment (zhuxiaoshang) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/RowTimeRangeBoundedPrecedingFunction.java:97]: I have saw the discussions in FLINK-15431,i'll follow it.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/RowTimeRangeBoundedPrecedingFunction.java [hunk 4]
```diff
@@ -86,12 +90,25 @@
...
-			GeneratedAggsHandleFunction genAggsHandler,
-			LogicalType[] accTypes,
-			LogicalType[] inputFieldTypes,
-			long precedingOffset,
-			int rowTimeIdx) {
+		GeneratedAggsHandleFunction genAggsHandler,
+		LogicalType[] accTypes,
+		LogicalType[] inputFieldTypes,
+		long precedingOffset,
+		int rowTimeIdx) {
```
[the remark this answers is not in the document]
Comment (zhuxiaoshang) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/RowTimeRangeBoundedPrecedingFunction.java]: Sorry for my IDEA code format changed this,i'll revert it.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/over/RowTimeOverWindowTestBase.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		new BigIntType(),
+		new BigIntType()
+	};
+	protected LogicalType[] accTypes = new LogicalType[]{new BigIntType()};
+
+	protected BinaryRowDataKeySelector keySelector = new BinaryRowDataKeySelector(new int[]{0}, inputFieldTypes);
+	protected TypeInformation<RowData> keyType = keySelector.getProducedType();
+
+	protected OneInputStreamOperatorTestHarness<RowData, RowData> createTestHarness(
+		KeyedProcessOperator<RowData, RowData, RowData> operator) throws Exception {
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/over/RowTimeOverWindowTestBase.java:56]:
```suggestion
			KeyedProcessOperator<RowData, RowData, RowData> operator) throws Exception {
```

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java [hunk 6]
```diff
@@ -68,6 +70,17 @@
 
 	protected transient AggsHandleFunction function;
 
+	// ------------------------------------------------------------------------
+	// Metrics
+	// ------------------------------------------------------------------------
+	protected static final String LATE_ELEMENTS_DROPPED_METRIC_NAME = "numLateRecordsDropped";
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java:76]: could be `private`?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java [hunk 7]
```diff
@@ -68,6 +70,17 @@
...
+	private transient Counter numLateRecordsDropped;
+
+	@VisibleForTesting
+	public Counter getCounter() {
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java:80]: could be `protected`?

**Review by libenchao — COMMENTED** [review]
@zhuxiaoshang Thanks for your updating, I added a few minor comments, it's better to address them before we merge this PR.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java (hunk 7 above)
Comment (zhuxiaoshang) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java:80]: Reason for consideration of future sub-class extension?In that way,i'll change the constant to `private` in `RowTimeRangeBoundedPrecedingFunction` and  `RowTimeRowsBoundedPrecedingFunction`.
WDYT

Comment (libenchao) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java:80]: we could not access it if it's `private`, `protected` methods can be accessed in tests.

Comment (zhuxiaoshang) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java:80]: we can access it by `getCounter()`

Comment (libenchao) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java:80]: I mean `protected getCounter()`

Comment (zhuxiaoshang) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractRowTimeUnboundedPrecedingOver.java:80]: got it

**Review by libenchao — APPROVED** [review]
@zhuxiaoshang Thanks for the update, the changes LGTM now, will merge.

**Comment by zhuxiaoshang** [comment]
@libenchao Why this PR is closed?

**Comment by libenchao** [comment]
@zhuxiaoshang I merged this manually, you can see it here: https://github.com/apache/flink/commit/b5b30650291e470a9635aef5f1348690bba78b29

**Pull request closed without merging**